src/Builder/Type/CombinedFieldQuery.php

```diff
@@ -4,28 +4,80 @@
 class CombinedFieldQuery implements FieldQueryInterface
 {
-    public function __construct(
-        /** @var list<FieldQueryInterface|Serializable|array|stdClass> $fieldQueries */
-        public readonly array $fieldQueries,
-    ) {
-        foreach ($fieldQueries as $fieldQuery) {
-            if (! $fieldQuery instanceof FieldQueryInterface && ! $fieldQuery instanceof Serializable && ! is_array($fieldQuery) && ! $fieldQuery instanceof stdClass) {
-                throw new InvalidArgumentException(sprintf('Expected filters to be a list of %s, %s, array or stdClass, %s given.', FieldQueryInterface::class, Document::class, get_debug_type($fieldQuery)));
+    /** @var list<FieldQueryInterface|array|stdClass> $fieldQueries */
+    public readonly array $fieldQueries;
+
+    /** @param list<FieldQueryInterface|array|stdClass> $fieldQueries */
+    public function __construct(array $fieldQueries)
+    {
+        if (! array_is_list($fieldQueries)) {
+            throw new InvalidArgumentException('Expected filters to be a list, invalid array given.');
+        }
+
+        // Flatten nested CombinedFieldQuery
+        $this->fieldQueries = array_reduce($fieldQueries, static function (array $fieldQueries, mixed $fieldQuery): array {
+            if ($fieldQuery instanceof CombinedFieldQuery) {
+                return array_merge($fieldQueries, $fieldQuery->fieldQueries);
+            }
+
+            $fieldQueries[] = $fieldQuery;
+
+            return $fieldQueries;
+        }, []);
+
+        // Validate FieldQuery types and non-duplicate operators
+        $seenOperators = [];
+        foreach ($this->fieldQueries as $fieldQuery) {
+            if ($fieldQuery instanceof FieldQueryInterface && $fieldQuery instanceof OperatorInterface) {
+                $operator = $fieldQuery->getOperator();
+            } elseif (is_array($fieldQuery)) {
+                if (count($fieldQuery) !== 1) {
+                    throw new InvalidArgumentException(sprintf('Operator array must contain exactly one key, %d given', count($fieldQuery)));
+                }
+
+                $operator = array_key_first($fieldQuery);
+                if (! str_starts_with($operator, '$')) {
+                    throw new InvalidArgumentException(sprintf('Operator array must contain exactly one key starting with $. "%s" given', $operator));
+                }
+            } elseif ($fieldQuery instanceof stdClass) {
+                $fieldQuery = get_object_vars($fieldQuery);
+                if (count($fieldQuery) !== 1) {
+                    throw new InvalidArgumentException(sprintf('Operator object must contain exactly one key. %d given', count($fieldQuery)));
+                }
+
+                $operator = array_key_first($fieldQuery);
+                if (! str_starts_with($operator, '$')) {
+                    throw new InvalidArgumentException(sprintf('Operator object must contain exactly one key starting with $. "%s" given', $operator));
+                }
```

Comment:
Suggestion: check for `stdClass` first and convert it to an array, which would allow you to use the same logic for both types.

Reply:
Done.